numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -1549,6 +1549,19 @@ gentype_@name@(PyObject *self, PyObject *args, PyObject *kwds)
 }
 /**end repeat**/
 
+static PyObject *
+gentype_dunder_round(PyObject *self, PyObject *args, PyObject *kwds)
+{
+    PyObject *obj = gentype_round(self, args, kwds);
+    if (obj == NULL) return NULL;
+
+    if (PyTuple_Size(args) == 0 &&
+        (kwds == NULL || PyDict_Size(kwds) == 0))
+        return PyNumber_Long(obj);
+    else
+        return PyNumber_Float(obj);
```

Comment:
This will downcast a `float128`, etc. which is incorrect. The result should be either a python integer or the old type. The second branch should thus not do anything

Furthermore, you need to go to a python integer for all of the following incarnations:
  * `round(scalar)` (done here)
  * `round(scalar, None)`
  * `round(scalar, ndigits=None)`

On the nitpicking front: Please add all braces, NumPy never leaves them out (I would probably be OK when it is on one line if someone wants to argue for it, but we currently are pretty strict on that style).

Reply:
> This will downcast a float128, etc. which is incorrect. The result should be either a python integer or the old type. The second branch should thus not do anything

This would produce very inconsistent results: Python scalars in one case and NumPy scalars in another case. I'd prefer to stick with one, either one is fine by me.

Reply:
Sorry, but to be clear, I do not think it is inconsistent result is is simply the *correct* result. `round` must be able to be a no-op given enough precision! The behaviour is only special for "round to integer" (even there I think it would be fine to return a non-python integer in principle, but python integers are awesome. Maybe I should have put this example immediately:
```
import decimal
round(decimal.Decimal("3.14"), 0)
# Decimal("3")
```
Everything else will basically destroy any precision we have.

Another thing we may have to think about are complex numbers, `round` is not defined for python complex, but we do define it. Maybe we should not implement `__round__`.

Reply:
We already implement it, removing it would be a huge break.

Reply:
Fair point about complex. We would have to deprecate the `dnigits=None` version. This now still adds the cast to integer, which for NumPy scalars, will give a warning and discard the imaginary part, while it gave a "reasonable" result before.

Reply:
I don't think a deprecation is necessary given this is the builtin `round` and not `np.round`. I would be happy to move the `__round__` method to `np.integer` and `np.floating`, and remove it from all others, in another PR.

Reply:
I don't want to slow you down, but, we have to be clear on what we want here.

1. Return the original type if `ndigits` is given (not `None`)
2. This currently, in theory, breaks working complex code by losing the complex part of the number (with a warning). A Deprecation to get rid of the `ndigits=None`, or `__round__` completely seems actually be more gentle to me.
3. I am not quite sure if `round(integer_type)` should not just keep returning the original integer type.

We decided to be bold enough to switch over to a (python) integer result for the default round. And yes it is new, but new is still many years old. I do not think we have actually made the decisions on those three points (or maybe I am just not aware). And my feeling is:

1. Absolutely, I am -10 on anything that casts to `float`.
2. My gut feeling is just deprecating is probably simpler.
3. I think returning the original integer type is probably just as well or better.

Which means in my view it would make more sense to limit this PR to `np.floating` and follow up with "whatever to do with complex" rather than the other way around.

Reply:
OTOH, I am not sure about point 2. maybe just returning a Python integer (similar to an `__index__()` call) is just as well. I am not sure I care either way much. I think it would be OK to return something as long as it supports `__index__()` and is an `numbers.Integral` but maybe there is no reason for _not_ converting it to a python int.

Reply:
```python
>>> round(5, 0)
5.0
```

So returning the original integer type would be a violation of the API. I'd be okay special-casing `np.inexact` to return the original type and also all others to return `np.float64`, but I'm unsure how I'd make that change.

Reply:
@hameerabbasi which python version are you using? On my python 3.7 and 3.8 that returns an integer, I think that must have been a bug in Python?

Reply:
Wait, you're right... I muscle-memory typed `python` without activating my environment, my bad.